### DISCUSSION TITLE
What is the best practice of protecting .env and the keys in it? [2022]

### DISCUSSION THREAD

**Discussion by Solemensis** [supabase/supabase#9151] [Questions] [answered]
Let's say i want to host my website on Netlify or Github Pages. But, my website uses an api with an api key (which is my Supabase api key in this example). So let's say i put that api key into .env file and run the "npm run build" command to convert my whole project into a ready to host dist/output folder (in vue). And hosted the /dist folder. I know, after using "npm run build" command, all the project being converted into a 1 small folder and .env(so my api keys) is not seen in the directory. But the question is, can client side people still see my api key from source code, or developer console, etc?

**In short, i'm searching for the standart way of securing these api keys when hosting the website.** Is going and hosting that dist/output folder enough for safety? If not, how web developers handle the secret token/api key securities when hosting a website(dist/output folder) with api's?

**Comment by Ziinc** [comment, marked as answer]
When distrbuting your supabase application, you should only include the `anon_key` and not the `service_role` key. the anon key is safe to expose to public, as there will be additional authentication required for protected data based on your row level security policies (so remember to configure them!)

When sending bundles to end users, one should expect api keys to be exposed, hence the use of jwt tokens and session keys so that there is another layer of auth. The exposed api keys only protect against non-client use of the api endpoint.
